File: Leetcode/Algorithm/cpp/00232-Implement Queue using Stacks.cc
class Queue {
public:
    Queue(): ptr(0) {}
    // Push element x to the back of queue.
    void push(int x) {
        st[ptr].push(x);
    }

    // Removes the element from in front of queue.
    void pop(void) {
        adjust();
        st[ptr^1].pop();
    }

    // Get the front element.
    int peek(void) {
        adjust();
        return st[ptr^1].top();
    }

    // Return whether the queue is empty.
    bool empty(void) {
        return st[0].empty() && st[1].empty();
    }
private:
    void adjust() {
        if (st[ptr^1].empty()) {
            int n = st[ptr].size();
            for (int i = 0; i < n; i++) {
                st[ptr^1].push(st[ptr].top());
                st[ptr].pop();
            }
        }
    }
private:
    stack<int> st[2];
    int ptr;
};

File: Leetcode/Algorithm/python/1000/00232-Implement Queue using Stacks.py
class Queue(object):
    def __init__(self):
        self.a = []
        self.b = []

    def push(self, x):
        self.a.append(x)

    def pop(self):
        if not self.b:
            self.do_move()
        return self.b.pop()
            
    def peek(self):
        if not self.b:
            self.do_move()
        return self.b[-1]

    def empty(self):
        return not self.a and not self.b
        
    def do_move(self):
        while self.a:
            self.b.append(self.a.pop())
        
